src/DGtal/geometry/volumes/distance/LpMetric.h

```diff
@@ -0,0 +1,237 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+#pragma once
+
+/**
+ * @file LpMetric.h
+ * @author David Coeurjolly (\c david.coeurjolly@liris.cnrs.fr )
+ * Laboratoire d'InfoRmatique en Image et Systèmes d'information - LIRIS (CNRS, UMR 5205), CNRS, France
+ *
+ * @date 2019/27/01
+ *
+ * Header file for module LpMetric.cpp
+ *
+ * This file is part of the DGtal library.
+ */
+
+#if defined(LpMetric_RECURSES)
+#error Recursive header files inclusion detected in LpMetric.h
+#else // defined(LpMetric_RECURSES)
+/** Prevents recursive inclusion of headers. */
+#define LpMetric_RECURSES
+
+#if !defined LpMetric_h
+/** Prevents repeated inclusion of headers. */
+#define LpMetric_h
+
+//////////////////////////////////////////////////////////////////////////////
+// Inclusions
+#include <iostream>
+#include <cmath>
+#include "DGtal/base/Common.h"
+#include "DGtal/kernel/CSpace.h"
+#include "DGtal/kernel/CEuclideanRing.h"
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+
+/////////////////////////////////////////////////////////////////////////////
+// template class LpMetric
+/**
+ * Description of template class 'LpMetric' <p>
+ * \brief Aim: implements  l_p metrics.
+ *
+ * Given a parameter p, the class implement classical l_p
+ * metric as a model of CMetricSpace. Hence, given two points
+ * @f$ x=(x_0...x_{n-1})@f$, @f$ y=(y_0...y_{n-1})@f$, we define a metric as:
+ *
+ * @f$ distance(x,y)= \left(
+ * \sum_{i=0}^{n-1} |x_i-y_i|^p\right)^{1/p}@f$
+ *
+ * This class performs all computations on C++ double converting the digital
+ * points to Space::RealPoint
+ *
+ * The exponent @a p is specifed at the constructor.
+ *
+ * @tparam TSpace the model of CSpace on which the metric is
+ * defined.
+ */
+  template <typename TSpace>
+  class LpMetric
+  {
+    // ----------------------- Standard services ------------------------------
+  public:
+
+    ///Copy the space type
+    typedef TSpace Space;
+    BOOST_CONCEPT_ASSERT(( concepts::CSpace<TSpace> ));
+    
+    ///Type for points (RealPoint for this class)
+    typedef typename Space::RealPoint Point;
+    ///Type for distance values
+    typedef double Value;
+    ///Type for raw distance values
+    typedef double RawValue;
+    
+    /**
+     * Default Constructor (p=2.0)
```

Comment:
I'm not completely comfortable with having a default constructor. When looking at code like
```C++
LpMetric<Space> metric;
```
I found that strange to not specify the `p`  and will have to look at the documentation to check the default behavior. I prefer to remove the default constructor and to be forced to always specify the power:
```C++
LpMetric<Space> metric(2);
```
Don't you?

Reply:
Agree, this would be clearer

Reply:
done